File: ios/MapMarkerManager.swift
import GoogleMaps
import UIKit

private final class MarkerState {
  var current: RNMarker
  var marker: GMSMarker?
  var currentToken: UInt64 = 0
  var appliedIcon: UIImage?
  var iconReady = false
  var anchorsDeferred = false
  var appliedStyleHash: NSNumber?
  var renderingStyleHash: NSNumber?
  var renderTask: Task<Void, Never>?
  var configurationErrorReported = false

  init(current: RNMarker) {
    self.current = current
  }
}

final class MapMarkerManager {
  private let builder: MapMarkerBuilder
  private let mapErrorHandler: MapErrorHandler
  private weak var mapView: GMSMapView?
  private var hasMapId = false
  private var advancedMarkersAvailable = false
  private var states: [String: MarkerState] = [:]
  private var iconGeneration: UInt64 = 0
  private var destroyed = false
  private var shownInfoWindowView: UIImageView?
  private var shownInfoWindowId: String?

  init(builder: MapMarkerBuilder, mapErrorHandler: MapErrorHandler) {
    self.builder = builder
    self.mapErrorHandler = mapErrorHandler
  }

  func attachMap(_ mapView: GMSMapView, hasMapId: Bool) {
    onMain {
      guard !self.destroyed else { return }
      self.mapView = mapView
      self.hasMapId = hasMapId
      self.advancedMarkersAvailable = mapView.mapCapabilities.contains(.advancedMarkers)
      self.states.values
        .filter { $0.marker == nil && $0.iconReady && $0.renderTask == nil }
        .forEach { self.addToMap($0) }
    }
  }

  func add(_ marker: RNMarker) {
    onMain {
      guard !self.destroyed else { return }
      self.remove(id: marker.id)
      let state = MarkerState(current: marker)
      self.states[marker.id] = state
      self.requestIcon(for: state)
    }
  }

  func update(_ next: RNMarker) {
    onMain {
      guard !self.destroyed else { return }
      guard let state = self.states[next.id] else { return }
      let prev = state.current
      if prev.markerEquals(next) { return }
      state.current = next

      if !prev.advancedEquals(next) {
        self.recreate(state)
        return
      }

      let nextStyleHash: NSNumber? = next.iconSvg != nil ? next.styleHash() : nil
      let renderingSameStyle = state.renderTask != nil && state.renderingStyleHash == nextStyleHash
      let iconUpToDate = state.renderTask == nil && state.iconReady && state.appliedStyleHash == nextStyleHash
      let needsRender = !renderingSameStyle && !iconUpToDate
      let deferAnchors = needsRender || state.renderTask != nil
      if deferAnchors, !prev.anchorEquals(next) || !prev.infoWindowAnchorEquals(next) {
        state.anchorsDeferred = true
      }

      state.marker.map { marker in
        self.builder.update(prev, next, marker, deferAnchors: deferAnchors)
        guard let mapView = self.mapView, mapView.selectedMarker == marker,
              !prev.infoWindowContentEquals(next) else { return }

        if next.infoWindowIsEmpty() {
          self.hideInfoWindow(id: next.id)
        } else if let infoWindowView = self.shownInfoWindowView,
                  self.shownInfoWindowId == next.id,
                  let updated = self.builder.buildInfoWindow(markerTag: marker.tagData) {
          infoWindowView.frame = updated.frame
          infoWindowView.image = updated.image
        } else if prev.infoWindowIconSvg != nil || next.infoWindowIconSvg != nil {
          self.showInfoWindow(id: next.id)
        }
      }

      if needsRender { self.requestIcon(for: state) }
    }
  }

  func remove(id: String) {
    onMain {
      self.states.removeValue(forKey: id).map { self.removeFromMap($0) }
    }
  }

  func showInfoWindow(id: String) {
    onMain {
      guard let mapView = self.mapView,
            let marker = self.states[id]?.marker else { return }
      mapView.selectedMarker = marker
    }
  }

  func hideInfoWindow(id: String) {
    onMain {
      guard let mapView = self.mapView,
            let marker = self.states[id]?.marker else { return }
      if mapView.selectedMarker == marker {
        mapView.selectedMarker = nil
      }
    }
  }

  func infoWindowView(markerTag: MarkerTag) -> UIImageView? {
    let view = builder.buildInfoWindow(markerTag: markerTag)
    shownInfoWindowView = view
    shownInfoWindowId = view != nil ? markerTag.id : nil
    setTracksInfoWindowChanges(id: markerTag.id, track: true)
    return view
  }

  func setTracksInfoWindowChanges(id: String, track: Bool) {
    onMain {
      guard let marker = self.states[id]?.marker,
            marker.tracksInfoWindowChanges != track else { return }
      marker.tracksInfoWindowChanges = track
    }
  }

  func clearIconCache() {
    builder.clearIconCache()
  }

  func updateAdvancedMarkersAvailable(_ available: Bool) {
    onMain {
      guard self.advancedMarkersAvailable != available else { return }
      self.advancedMarkersAvailable = available
      self.states.values
        .filter { $0.current.usesAdvancedMarker() }
        .forEach { self.recreate($0) }
    }
  }

  func destroy() {
    onMain {
      self.destroyed = true
      self.states.values.forEach { self.removeFromMap($0) }
      self.states.removeAll()
      self.builder.clearIconCache()
      self.mapView = nil
      CATransaction.flush()
    }
  }

  private func requestIcon(for state: MarkerState) {
    state.renderTask?.cancel()
    state.renderTask = nil
    iconGeneration += 1
    let token = iconGeneration
    state.currentToken = token
    let id = state.current.id

    guard let iconSvg = state.current.iconSvg else {
      state.renderingStyleHash = nil
      applyIcon(id: id, token: token, icon: nil)
      return
    }

    let styleHash = state.current.styleHash()
    if let cached = builder.cachedIcon(styleHash: styleHash) {
      state.renderingStyleHash = nil
      applyIcon(id: id, token: token, icon: cached)
      return
    }

    state.renderingStyleHash = styleHash
    state.renderTask = builder.renderIcon(
      markerId: id,
      iconSvg: iconSvg,
      styleHash: styleHash
    ) { [weak self] icon in
      self?.applyIcon(id: id, token: token, icon: icon)
    }
  }

  private func applyIcon(id: String, token: UInt64, icon: UIImage?) {
    guard let state = states[id] else { return }
    guard state.currentToken == token else { return }
    state.renderTask = nil
    state.renderingStyleHash = nil
    state.appliedStyleHash = state.current.iconSvg != nil ? state.current.styleHash() : nil
    state.iconReady = true

    if let marker = state.marker {
      withCATransaction(disableActions: true) {
        marker.tracksViewChanges = true
        marker.icon = icon
        if state.anchorsDeferred {
          self.builder.applyAnchors(state.current, marker)
          state.anchorsDeferred = false
        }
        marker.tracksViewChanges = false
      }
      return
    }

    state.appliedIcon = icon
    if mapView != nil {
      addToMap(state)
    }
  }

  private func addToMap(_ state: MarkerState) {
    if let message = state.current.advancedMarkerConfigurationError() {
      if !state.configurationErrorReported {
        state.configurationErrorReported = true
        mapErrorHandler.report(
          RNMapErrorCode.invalidArgument,
          "markerId=\(state.current.id) \(message)"
        )
      }
      state.appliedIcon = nil
      return
    }

    if state.current.usesAdvancedMarker(), !hasMapId {
      if !state.configurationErrorReported {
        state.configurationErrorReported = true
        mapErrorHandler.report(
          RNMapErrorCode.invalidArgument,
          "markerId=\(state.current.id) Advanced Markers require initialProps.mapId"
        )
      }
      state.appliedIcon = nil
      return
    }

    let useAdvancedMarker = state.current.usesAdvancedMarker() && advancedMarkersAvailable
    let marker = builder.build(
      state.current,
      icon: state.appliedIcon,
      useAdvancedMarker: useAdvancedMarker
    )
    marker.map = mapView
    state.marker = marker
    state.appliedIcon = nil
    state.anchorsDeferred = false
  }

  private func removeFromMap(_ state: MarkerState) {
    state.renderTask?.cancel()
    if shownInfoWindowId == state.current.id {
      shownInfoWindowView = nil
      shownInfoWindowId = nil
    }
    state.marker.map {
      $0.icon = nil
      $0.map = nil
    }
    state.marker = nil
  }

  private func recreate(_ state: MarkerState) {
    removeFromMap(state)
    state.iconReady = false
    state.appliedIcon = nil
    state.appliedStyleHash = nil
    state.anchorsDeferred = false
    state.configurationErrorReported = false
    requestIcon(for: state)
  }
}

File: android/src/main/java/com/rngooglemapsplus/MapMarkerManager.kt
package com.rngooglemapsplus

import android.widget.ImageView
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.BitmapDescriptor
import com.google.android.gms.maps.model.Marker
import com.rngooglemapsplus.extensions.anchorEquals
import com.rngooglemapsplus.extensions.infoWindowAnchorEquals
import com.rngooglemapsplus.extensions.infoWindowContentEquals
import com.rngooglemapsplus.extensions.infoWindowIsEmpty
import com.rngooglemapsplus.extensions.markerEquals
import com.rngooglemapsplus.extensions.styleHash
import com.rngooglemapsplus.extensions.toMarkerTag
import kotlinx.coroutines.Job

private class MarkerState(
  var current: RNMarker,
) {
  var marker: Marker? = null
  var currentToken: Long = 0L
  var appliedIcon: BitmapDescriptor? = null
  var appliedHitbox: MarkerIconHitbox? = null
  var iconReady: Boolean = false
  var anchorsDeferred: Boolean = false
  var appliedStyleHash: Int? = null
  var renderingStyleHash: Int? = null
  var renderJob: Job? = null
}

class MapMarkerManager(
  private val builder: MapMarkerBuilder,
) {
  private var map: GoogleMap? = null
  private val states = mutableMapOf<String, MarkerState>()
  private var iconGeneration = 0L
  private var destroyed = false

  fun attachMap(map: GoogleMap) =
    onUi {
      if (destroyed) return@onUi
      this.map = map
      states.values
        .filter { it.marker == null && it.iconReady && it.renderJob == null }
        .forEach { addToMap(it) }
    }

  fun add(marker: RNMarker) =
    onUi {
      if (destroyed) return@onUi
      remove(marker.id)
      val state = MarkerState(marker)
      states[marker.id] = state
      requestIcon(state)
    }

  fun update(next: RNMarker) =
    onUi {
      if (destroyed) return@onUi
      val state = states[next.id] ?: return@onUi
      val prev = state.current
      if (prev.markerEquals(next)) return@onUi
      state.current = next

      val nextStyleHash = if (next.iconSvg != null) next.styleHash() else null
      val renderingSameStyle = state.renderJob != null && state.renderingStyleHash == nextStyleHash
      val iconUpToDate = state.renderJob == null && state.iconReady && state.appliedStyleHash == nextStyleHash
      val needsRender = !renderingSameStyle && !iconUpToDate
      val deferAnchors = needsRender || state.renderJob != null
      if (deferAnchors && (!prev.anchorEquals(next) || !prev.infoWindowAnchorEquals(next))) {
        state.anchorsDeferred = true
      }

      state.marker?.let { marker ->
        builder.update(prev, next, marker, deferAnchors)
        val infoWindowNeedsRefresh =
          !prev.infoWindowContentEquals(next) ||
            !prev.infoWindowAnchorEquals(next) ||
            prev.rotation != next.rotation
        if (marker.isInfoWindowShown && infoWindowNeedsRefresh) {
          if (next.infoWindowIsEmpty()) {
            hideInfoWindow(next.id)
          } else {
            showInfoWindow(next.id)
          }
        }
      }

      if (needsRender) requestIcon(state)
    }

  fun remove(id: String) =
    onUi {
      states.remove(id)?.let { removeFromMap(it) }
    }

  fun showInfoWindow(id: String) =
    onUi {
      states[id]?.marker?.showInfoWindow()
    }

  fun hideInfoWindow(id: String) =
    onUi {
      states[id]?.marker?.hideInfoWindow()
    }

  fun infoWindowView(markerTag: MarkerTag): ImageView? = builder.buildInfoWindow(markerTag)

  fun clearIconCache() = builder.clearIconCache()

  fun destroy() =
    onUi {
      destroyed = true
      states.values.forEach { removeFromMap(it) }
      states.clear()
      builder.clearIconCache()
      map = null
    }

  private fun requestIcon(state: MarkerState) {
    state.renderJob?.cancel()
    state.renderJob = null
    iconGeneration += 1
    val token = iconGeneration
    state.currentToken = token
    val id = state.current.id

    val iconSvg = state.current.iconSvg
    if (iconSvg == null) {
      state.renderingStyleHash = null
      applyIcon(id, token, null, null)
      return
    }

    val styleHash = state.current.styleHash()
    builder.cachedIcon(styleHash)?.let { cached ->
      state.renderingStyleHash = null
      applyIcon(id, token, cached, builder.buildHitbox(iconSvg))
      return
    }

    state.renderingStyleHash = styleHash
    state.renderJob =
      builder.renderIcon(id, iconSvg, styleHash) { icon, hitbox ->
        applyIcon(id, token, icon, hitbox)
      }
  }

  private fun applyIcon(
    id: String,
    token: Long,
    icon: BitmapDescriptor?,
    hitbox: MarkerIconHitbox?,
  ) {
    val state = states[id] ?: return
    if (state.currentToken != token) return
    state.renderJob = null
    state.renderingStyleHash = null
    state.appliedStyleHash = if (state.current.iconSvg != null) state.current.styleHash() else null
    state.iconReady = true

    val marker = state.marker
    if (marker != null) {
      marker.setIcon(icon)
      if (state.anchorsDeferred) {
        builder.applyAnchors(state.current, marker)
        state.anchorsDeferred = false
      }
      marker.tag = state.current.toMarkerTag(hitbox)
      return
    }

    state.appliedIcon = icon
    state.appliedHitbox = hitbox
    if (map != null) addToMap(state)
  }

  private fun addToMap(state: MarkerState) {
    state.marker =
      map?.addMarker(builder.build(state.current, state.appliedIcon))?.apply {
        tag = state.current.toMarkerTag(state.appliedHitbox)
      }
    state.appliedIcon = null
    state.appliedHitbox = null
    state.anchorsDeferred = false
  }

  private fun removeFromMap(state: MarkerState) {
    state.renderJob?.cancel()
    state.marker?.remove()
  }
}
